addons/mail/models/ir_attachment.py

```diff
@@ -34,3 +34,15 @@ def register_as_main_attachment(self, force=True):
                     related_record.message_main_attachment_id = self
                 except AccessError:
                     pass
+
+    def _attachment_format(self, main_attachment = None, safari = None):
+        return [{
```

Comment:
For performance and override reasons it would be better if each formatter returned the result as a dict (with record as key) instead of a list. It makes it much easier to find back an element on the result, and prevents potential inconsistencies of having the same record multiple times in the array.